src/ffi/logging.rs

```diff
@@ -0,0 +1,232 @@
+            self.cb
+                .log(self.level, filtered_msg)
+                .map_err(io::Error::other)?;
+        }
+
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+struct LogStatus {
+    string: String,
+    counter: u32,
+}
+
+lazy_static::lazy_static! {
+    static ref LAST_LOG_STATUS: Mutex<LogStatus> = {
+        Mutex::new(LogStatus{string: String::default(), counter: 0})
+    };
+}
+
+fn filter_log_message(msg: String) -> Option<String> {
+    let mut log_status = match LAST_LOG_STATUS.lock() {
+        Ok(status) => status,
+        Err(_) => {
+            return None;
+        }
+    };
+
+    if !log_status.string.eq(&msg) {
+        log_status.string = msg.clone();
+        log_status.counter = 0;
+        return Some(msg);
+    }
+
+    if log_status.counter > 0 && log_status.counter % 100 == 0 {
+        log_status.counter += 1;
+        return Some(format!("[repeated 100 times!] {}", msg));
+    }
+
+    if log_status.counter < 10 {
+        log_status.counter += 1;
+        return Some(msg);
+    }
+
+    log_status.counter += 1;
+    None
+}
+
+#[cfg(test)]
+mod test {
+
+    /// Added in external crate to avoid line changes as much as possible
+    use tracing::{debug, info, trace, warn};
+
+    use super::*;
+
+    use std::{
+        fmt::Debug,
+        sync::{Arc, Mutex},
+    };
+
+    use crate::{TelioLogLevel, TelioLoggerCb};
+
+    #[test]
+    fn test_trace_via_telio_cb() {
+        let start = line!() + 1;
+        let act = || {
+            trace!("first message"); // +1
+            debug!("second message"); // +2
+            info!("third\nmutiline\nmessage"); // +3
+            warn!(
+                n = 2,
+                extra = "extra info",
+                "fourth message with {}",
+                "info"
+            ); // +4
+        };
+        let mpath = module_path!();
+        let expected = [
+            (
+                TelioLogLevel::Debug,
+                format!("{:?}:{} second message", mpath, start + 2),
+            ),
+            (
+                TelioLogLevel::Info,
+                format!("{:?}:{} third\nmutiline\nmessage", mpath, start + 3),
+            ),
+            (
+                TelioLogLevel::Warning,
+                format!(
+                    "{:?}:{} fourth message with info n=2 extra=\"extra info\"",
+                    mpath,
+                    start + 4
+                ),
+            ),
+        ];
+        let logs = Log::default();
+        let sub = build(TelioLogLevel::Debug, Box::new(logs.clone()));
+
+        tracing::subscriber::with_default(sub, act);
+
+        let actual = logs.0.lock().unwrap().clone();
+
+        assert_eq!(&expected[..], &actual[..])
+    }
+
+    #[derive(Default, Clone, Debug)]
+    struct Log(Arc<Mutex<Vec<(TelioLogLevel, String)>>>);
+    impl TelioLoggerCb for Log {
+        fn log(&self, level: TelioLogLevel, payload: String) -> crate::FFIResult<()> {
```

Comment:
Hmm, it is still `FFI` instead of `Ffi`

Reply:
I see lsp did not include #[cfg(test)] and docs also :|